protocols/jvss/jvss.go

```diff
@@ -15,645 +21,249 @@ import (
-// poly.Threshold. You can give a new one after calling this function.
-func NewJVSSProtocol(n *sda.Node) (*JVSSProtocol, error) {
-	// find ourselves in the entityList
-	var idx int = -1
-	// at the same time create the public list
-	tree := n.Tree()
-	nodes := tree.List()
-	pubs := make([]abstract.Point, len(nodes))
+// JVSS is the main protocol struct and implements the sda.ProtocolInstance
+// interface.
+type JVSS struct {
+	*sda.Node                         // The SDA TreeNode
+	keyPair     *config.KeyPair       // KeyPair of the host
+	nodeList    []*sda.TreeNode       // List of TreeNodes in the JVSS group
+	pubKeys     []abstract.Point      // List of public keys of the above TreeNodes
+	info        poly.Threshold        // JVSS thresholds
+	schnorr     *poly.Schnorr         // Long-term Schnorr struct to compute distributed signatures
+	secrets     map[string]*Secret    // Shared secrets (long- and short-term ones)
+	ltssInit    bool                  // Indicator whether shared secret has been already initialised or not
+	secretsDone chan bool             // Channel to indicate when shared secrets of all peers are ready
+	sigChan     chan *poly.SchnorrSig // Channel for JVSS signature
+}
+
+// Secret contains all information for long- and short-term shared secrets.
+type Secret struct {
+	secret   *poly.SharedSecret              // Shared secret
+	receiver *poly.Receiver                  // Receiver to aggregate deals
+	deals    map[int]*poly.Deal              // Buffer for deals
+	sigs     map[int]*poly.SchnorrPartialSig // Buffer for partial signatures
+	numConfs int                             // Number of collected confirmations that shared secrets are ready
+	mtx      *sync.Mutex                     // Mutex to sync access to numConfs
+}
+
+// NewJVSS creates a new JVSS protocol instance and returns it.
+func NewJVSS(node *sda.Node) (sda.ProtocolInstance, error) {
+
+	kp := &config.KeyPair{Suite: node.Suite(), Public: node.Public(), Secret: node.Private()}
+	nodes := node.Tree().List()
+	pk := make([]abstract.Point, len(nodes))
 	for i, tn := range nodes {
-		if tn.Id.Equals(n.TreeNode().Id) {
-			idx = i
+		pk[i] = tn.Entity.Public
+	}
+	// NOTE: T <= R <= N (for simplicity we use T = R = N; might change later)
+	info := poly.Threshold{T: len(nodes), R: len(nodes), N: len(nodes)}
+
+	jv := &JVSS{
+		Node:        node,
+		keyPair:     kp,
+		nodeList:    nodes,
+		pubKeys:     pk,
+		info:        info,
+		schnorr:     new(poly.Schnorr),
+		secrets:     make(map[string]*Secret),
+		ltssInit:    false,
+		secretsDone: make(chan bool, 1),
+		sigChan:     make(chan *poly.SchnorrSig),
+	}
+
+	// Setup message handlers
+	handlers := []interface{}{
+		jv.handleSecInit,
+		jv.handleSecConf,
+		jv.handleSigReq,
+		jv.handleSigResp,
+	}
+	for _, h := range handlers {
+		if err := jv.RegisterHandler(h); err != nil {
+			return nil, fmt.Errorf("Error, could not register handler: " + err.Error())
```

Comment:
Do we need to put that into the `Node`? Having a `Node.RegisterHandlers` with a varargs?